### ISSUE TITLE
Debugging source code

### ISSUE THREAD

**Issue by angriestalien** [NationalSecurityAgency/ghidra#518]
What steps should I complete to debug JarRun for AnalyzeHeadless or GhidraRun? I’m using a mac running eclipse. 

•	Are there eclipse settings I need to set to make application.properties file accessible?
•	How do I make the Configurations and Extensions folder accessible during debug?
•	Do I need to replace “/_Root/Ghidra/” with something else?

I was able to load the application.properites with a hack but was stuck again when ghidra was trying to load defaultTools.

The error is thrown because the path to application.properties is invalid.

Exception in thread "main" java.lang.NullPointerException
	at ghidra.GhidraJarApplicationLayout.findGhidraApplicationRootDirs(GhidraJarApplicationLayout.java:53)
	at ghidra.GhidraApplicationLayout.<init>(GhidraApplicationLayout.java:46)
	at ghidra.GhidraJarApplicationLayout.<init>(GhidraJarApplicationLayout.java:41)
	at ghidra.JarRun.main(JarRun.java:27)

Thanks for the help.

**Comment by ryanmkurtz** [comment]
How are you launching JarRun?  Is your eclipse environment just the git repo?

**Comment by angriestalien** [comment]
I'm launching JarRun with a Run Configuration in the Feature Base project. I pass these arguments: JarRun testing -import ~/binaries/cat

testing is a ghidra project I created with Ghidra 9.0.2.
cat is /bin/cat 

The environment is the git repo (commit 1cae5552). I followed the instructions in DevGuide.md.

**Comment by ryanmkurtz** [comment]
Do you need to launch with JarRun?  That is a very special case entry point into Ghidra.  What is the main thing you are trying to debug?  There are probably easier ways to do it.

**Comment by angriestalien** [comment]
I'm not too familiar with the code base. Still trying to learn it. I'll try to debug from another entry point if you can give me some direction. 

I want to extract the decompiled C from an AnalyzeHeadless launch.

**Comment by ryanmkurtz** [comment]
The only debugging we really provide out of the box is through the GUI with the Ghidra run configuration in Eclipse.  It's under the Java Application category in run configurations.  Just launch that in debug mode and any of your Java breakpoints will catch.  If you really need it to be a headless launch I can help you create that run configuration but going through the GUI is probably fine for what you need.

**Comment by angriestalien** [comment]
Thanks for the advice. I can experiment with the GUI. However, I think I will need to use the AnalyzeHeadless launch. If it is not too much trouble, I would really appreciate the help creating a run configuration for the headless launch.

**Comment by ryanmkurtz** [comment]
If you go into the Eclipse run configurations, click the button to duplicate the Ghidra run configuration.  Then in the arguments tab, change ghidra.GhidraRun to ghidra.app.util.headless.AnalyzeHeadless followed by the command line arguments you would normally pass into the analyzeHeadless script.

**Comment by ryanmkurtz** [comment]
Did that end up working for you?

**Comment by angriestalien** [comment]
Unfortunately, no. It throws a ClassNotFoundException.  ghidra.app.util.headless.AnalyzeHeadless is not in the Framework Utility project. Also, ghidra.app.util.headless.AnalyzeHeadless does not have a main class.

The Ghidra run configuration I have is with project set to Framework Utility and a main class of ghidra.GhidraLauncher. Maybe I have the wrong Ghidra run configuration?

**Comment by ryanmkurtz** [comment]
Ghidra uses it's own classloader to discover all the modules when it starts up, so the entry point is always the GhidraLauncher.  GhidraLauncher redirects execution to class specified by your first command line argument, passing along the rest of the command line arguments to the thing you are redirecting to, like the GUI (GhidraRun) or headless (AnalyzeHeadless).

**Comment by angriestalien** [comment]
Thank you. I was able to debug an AnalyzeHeadless launch. That will keep me going. 

I was thinking the AnalyzeHeadless run would allow me to analyze a directory of binaries to view in the gui later. I opened the project I created by running AnalyzeHeadless configuration and see the files AnalyzeHeadless analyzed. However, when I select the analyzed file and click the CodeBrowser, the program Trees pane displays a folder with text saying No Program. 

Am I missing an argument?

ghidra.app.util.headless.AnalyzeHeadless /Users/user1/ghidraTest newProject2 -import /Users/user1/binaries/cat
